---
title: "Gene ID Conversion data setup"
output: html_notebook
---


This notebook is to set up the reference data for gene ID conversion and ultimately deduplication.

```{r}
suppressPackageStartupMessages({
  library(SingleCellExperiment)
  library(dplyr)
})
```


Read in the test data ScPCA SCE object and extract the row data:

```{r}
genes_scpca <- readRDS(here::here("tests", "testthat", "data", "scpca_sce.rds")) |>
  rowData() |>
  as.data.frame() |>
  # Use Ensembl ID if gene symbol is missing, then make unique
  mutate(
    gene_symbol_scpca = ifelse(is.na(gene_symbol), gene_ids, gene_symbol),
    gene_symbol_scpca_unique = make.unique(gene_symbol_scpca)
  ) |>
  select(gene_ids, gene_symbol_scpca, gene_symbol_scpca_unique)
```


Download and read in a 2020 10x reference dataset and extract the gene symbols.
Note that the 2020 Cell Ranger reference does not use Ensembl gene IDs for missing symbols, but the 2024 reference does.

```{r}
# Download data
url_10x2020 <- "https://cf.10xgenomics.com/samples/cell-exp/7.0.1/SC3pv3_GEX_Human_PBMC/SC3pv3_GEX_Human_PBMC_filtered_feature_bc_matrix.h5" # nolint
temp_10x2020 <- tempfile(fileext = ".h5")
download.file(url_10x2020, temp_10x2020, mode = "wb")
on.exit(unlink(temp_10x2020), add = TRUE) # delete when done
```

```{r}
# Read in the data
genes_10x2020 <- DropletUtils::read10xCounts(temp_10x2020) |>
  rowData() |>
  as.data.frame() |>
  filter(Type == "Gene Expression") |>
  rename(
    gene_ids = ID,
    gene_symbol_10x2020 = Symbol
  ) |>
  # Use Ensembl ID if gene symbol is missing, then make unique
  mutate(gene_symbol_10x2020_unique = make.unique(gene_symbol_10x2020)) |>
  select(gene_ids, gene_symbol_10x2020, gene_symbol_10x2020_unique)
```
Download and read in a 2024 10x reference dataset and extract the gene symbols:

```{r}
# Download data
url_10x2024 <- "https://cf.10xgenomics.com/samples/cell-exp/9.0.0/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x/5k_Human_Donor2_PBMC_3p_gem-x_5k_Human_Donor2_PBMC_3p_gem-x_count_sample_filtered_feature_bc_matrix.h5" # nolint
temp_10x2024 <- tempfile(fileext = ".h5")
download.file(url_10x2024, temp_10x2024, mode = "wb")
on.exit(unlink(temp_10x2024), add = TRUE) # delete when done
```


```{r}
genes_10x2024 <- DropletUtils::read10xCounts(temp_10x2024) |>
  rowData() |>
  as.data.frame() |>
  filter(Type == "Gene Expression") |>
  rename(
    gene_ids = ID,
    gene_symbol_10x2024 = Symbol
  ) |>
  # Use Ensembl ID if gene symbol is missing, then make unique
  mutate(gene_symbol_10x2024_unique = make.unique(gene_symbol_10x2024)) |>
  select(gene_ids, gene_symbol_10x2024, gene_symbol_10x2024_unique)
```

## Merge the gene symbol tables

```{r}
scpca_gene_reference <- genes_scpca |>
  full_join(genes_10x2020, by = "gene_ids") |>
  full_join(genes_10x2024, by = "gene_ids")
```


## Look at some stats for the tables

We expect that many of the Ensembl IDs in the ScPCA data will not have corresponding values in the 10x references, but some of the 10x references have Ensembl IDs not present in the ScPCA data.
Let's look at the number of missing gene symbols in the ScPCA data:

```{r}
scpca_missing <- scpca_gene_reference |>
  filter(is.na(gene_symbol_scpca))
```

How many of the gene symbols in the 10x 2020 reference are in the ScPCA data, but perhaps with a different Ensembl ID?

```{r}
symbols_10x2020 <- scpca_missing$gene_symbol_10x2020[!is.na(scpca_missing$gene_symbol_10x2020)]
sum(symbols_10x2020 %in% genes_scpca$gene_symbol_scpca)
```

What are they?

```{r}
# Match the genes that are in the 10x 2020 reference to the ScPCA table on gene symbol
# Only for those where the gene id is missing in the ScPCA table
genes_10x2020 |>
  filter(
    gene_symbol_10x2020 %in% genes_scpca$gene_symbol_scpca,
    !gene_ids %in% genes_scpca$gene_ids
  ) |>
  left_join(genes_scpca, by = c("gene_symbol_10x2020" = "gene_symbol_scpca"))
```

Looks like these are all cases where the gene symbol has been updated.
In only one case is this a gene where something was not unique, which is `LINC01505`.
This gene seems to have simply been removed in later revisions, so I think we can safely not worry about it.

```{r}
scpca_gene_reference |>
  filter(gene_symbol_10x2020 == "LINC01505")
```

In general, simply translating to the list gene symbols should work as expected.

### BAC removal

In Ensembl version 104, the version we are using for ScPCA, BAC-based gene IDs were removed and replaced with simply the Ensembl gene ID.
Many of the gene symbols that are present in the 10x references but missing in the ScPCA data are these BAC-based gene IDs.
We will probably want to translate these to the 10x symbol when requested.

```{r}
scpca_gene_reference |>
  filter(
    !is.na(gene_symbol_10x2020),
    !gene_symbol_10x2020 %in% genes_scpca$gene_symbol_scpca
  )
```

### Disagreements between symbols

One other question is how often the `unique` gene symbols disagree for the same Ensemble ID.
Here we will exclude the cases where the gene symbol is an Ensembl ID in the ScPCA data, as these are expected to be different.


```{r}
scpca_gene_reference |>
  filter(
    gene_symbol_scpca != gene_ids,
    gene_symbol_scpca_unique != gene_symbol_10x2020_unique
  )
```

More than I expected, but it looks like most of these are cases where the gene symbol has been updated.
What we are really interested in is the cases where the process of making the gene symbol unique had different effects:

```{r}
scpca_gene_reference |>
  filter(
    gene_symbol_scpca == gene_symbol_10x2020,
    gene_symbol_scpca_unique != gene_symbol_10x2020_unique
  )
```

This definitely happens more than I would have hoped. Which is to say that the best way to handle this is probably simply to translate using the the table directly when performing comparisons to existing data for the most accurate results.

We will need to have clear instructions about when to use which kind of translation.

## Add the table to package data

```{r}
usethis::use_data(scpca_gene_reference, internal = TRUE, version = 3, overwrite = TRUE)
```
